File: historian/scripts/udp_sender.ts
import dgram from "node:dgram";
import { loadConfig } from "../src/config/loader";
import { encodeUdpBatch } from "../src/storage/codec";
import { Point, ValueTypeCode } from "../src/types/valueTypes";

const THREE_DAYS_SECONDS = 3 * 24 * 60 * 60;

function toEpochUnitFromMs(ms: bigint, unit: "us" | "ns"): bigint {
  return unit === "ns" ? ms * 1_000_000n : ms * 1_000n;
}

function buildThreeTagSecond(baseTs: bigint, seq: number): Point[] {
  const out: Point[] = [];
  out.push({
    tagId: 1,
    tsEpoch: baseTs,
    typeCode: ValueTypeCode.Int32,
    value: (seq % 2000) - 1000
  });
  out.push({
    tagId: 2,
    tsEpoch: baseTs,
    typeCode: ValueTypeCode.Float32,
    value: Math.sin(seq / 25) * 100
  });
  out.push({
    tagId: 3,
    tsEpoch: baseTs,
    typeCode: ValueTypeCode.String,
    value: `state-${seq % 5}`
  });
  return out;
}

async function main(): Promise<void> {
  const config = await loadConfig();
  const socket = dgram.createSocket("udp4");
  const secondsPerPacket = Number(process.argv[2] ?? 600);
  const intervalMs = Number(process.argv[3] ?? 0);
  const nowMs = BigInt(Date.now());
  const startMs = nowMs - 3n * 24n * 60n * 60n * 1000n;
  const startTs = toEpochUnitFromMs(startMs, config.storage.timestampUnit);
  const step = config.storage.timestampUnit === "ns" ? 1_000_000_000n : 1_000_000n;
  let sentPoints = 0;
  let packets = 0;

  // eslint-disable-next-line no-console
  console.log(
    `sending 3-day per-second data automatically: start=${new Date(Number(startMs)).toISOString()} end=${new Date(Number(nowMs)).toISOString()}`
  );

  for (let sec = 0; sec < THREE_DAYS_SECONDS; sec += secondsPerPacket) {
    const points: Point[] = [];
    const until = Math.min(THREE_DAYS_SECONDS, sec + secondsPerPacket);
    for (let s = sec; s < until; s++) {
      const ts = startTs + BigInt(s) * step;
      points.push(...buildThreeTagSecond(ts, s));
    }
    const payload = encodeUdpBatch(points);
    await new Promise<void>((resolve, reject) => {
      socket.send(payload, config.udp.port, config.udp.host, (err) => (err ? reject(err) : resolve()));
    });
    sentPoints += points.length;
    packets += 1;
    if (packets % 50 === 0) {
      const progress = ((until / THREE_DAYS_SECONDS) * 100).toFixed(1);
      // eslint-disable-next-line no-console
      console.log(`sent packets=${packets}, points=${sentPoints}, progress=${progress}%`);
    }
    if (intervalMs > 0) await new Promise((r) => setTimeout(r, intervalMs));
  }

  socket.close();
  // eslint-disable-next-line no-console
  console.log(`finished sending ${sentPoints} points (${THREE_DAYS_SECONDS} seconds x 3 tags)`);
}

void main().catch((err) => {
  // eslint-disable-next-line no-console
  console.error(err);
  process.exit(1);
});

File: historian_v2/src/bin/udp_sender.rs
use anyhow::Result;
use historian_v2::config::{load_config, TimestampUnit};
use historian_v2::storage::codec::encode_udp_batch;
use historian_v2::types::{Point, PointValue, ValueTypeCode};
use tokio::net::UdpSocket;

const THREE_DAYS_SECONDS: i64 = 3 * 24 * 60 * 60;

fn to_epoch(ms: i64, unit: &TimestampUnit) -> i64 {
    match unit {
        TimestampUnit::Us => ms * 1_000,
        TimestampUnit::Ns => ms * 1_000_000,
    }
}

fn per_second_points(ts: i64, seq: i64) -> Vec<Point> {
    vec![
        Point {
            tag_id: 1,
            ts_epoch: ts,
            type_code: ValueTypeCode::Int32,
            value: PointValue::Num((seq % 2000 - 1000) as f64),
        },
        Point {
            tag_id: 2,
            ts_epoch: ts,
            type_code: ValueTypeCode::Float32,
            value: PointValue::Num((seq as f64 / 25.0).sin() * 100.0),
        },
        Point {
            tag_id: 3,
            ts_epoch: ts,
            type_code: ValueTypeCode::String,
            value: PointValue::Str(format!("state-{}", seq % 5)),
        },
    ]
}

#[tokio::main]
async fn main() -> Result<()> {
    let cfg = load_config("historian.config.json").await?;
    let seconds_per_packet: i64 = std::env::args()
        .nth(1)
        .and_then(|v| v.parse().ok())
        .unwrap_or(600);
    let interval_ms: u64 = std::env::args()
        .nth(2)
        .and_then(|v| v.parse().ok())
        .unwrap_or(0);

    let now_ms = chrono::Utc::now().timestamp_millis();
    let start_ms = now_ms - 3 * 24 * 3600 * 1000;
    let start_ts = to_epoch(start_ms, &cfg.storage.timestamp_unit);
    let step = to_epoch(1000, &cfg.storage.timestamp_unit);
    let socket = UdpSocket::bind("0.0.0.0:0").await?;
    let addr = format!("{}:{}", cfg.udp.host, cfg.udp.port);
    let mut sent = 0usize;

    println!(
        "sending 3-day per-second data: {} -> {}",
        chrono::DateTime::<chrono::Utc>::from_timestamp_millis(start_ms).unwrap(),
        chrono::DateTime::<chrono::Utc>::from_timestamp_millis(now_ms).unwrap()
    );

    let mut sec = 0i64;
    while sec < THREE_DAYS_SECONDS {
        let until = (sec + seconds_per_packet).min(THREE_DAYS_SECONDS);
        let mut batch = Vec::new();
        for s in sec..until {
            let ts = start_ts + s * step;
            batch.extend(per_second_points(ts, s));
        }
        let payload = encode_udp_batch(&batch);
        socket.send_to(&payload, &addr).await?;
        sent += batch.len();
        sec = until;
        if interval_ms > 0 {
            tokio::time::sleep(std::time::Duration::from_millis(interval_ms)).await;
        }
    }
    println!("finished sending {} points", sent);
    Ok(())
}
